File: py/438.py
from typing import *
import string

class Solution:
    def is_equal(self, ms, mp):
        for c in string.ascii_lowercase:
            if ms[c] != mp[c]:
                return False 
        return True

    def findAnagrams(self, s: str, p: str) -> List[int]:
        ms = {}
        mp = {}
        
        for c in string.ascii_lowercase:
            ms[c] = 0
            mp[c] = 0

        for c in p:
            mp[c] += 1

        res = []
        l = 0
        for r in range(len(s)):
            ms[s[r]] += 1
            if r - l + 1 < len(p):
                continue
            if self.is_equal(ms, mp):
                res.append(l)
                
            ms[s[l]] -= 1
            l += 1

        return res



solution = Solution()

for s, p in [("cbaebabacd", "abc"), ("abab", "ab")]:
    result = solution.findAnagrams(s, p)
    print(result)

File: go/438.go
package main 


func findAnagrams(s string, p string) []int {
    ms := [26]byte{}
    mp := [26]byte{}
    
    for _, c := range p {
        mp[c - 'a'] += 1
    }
    
    res := []int{}
    l := 0 
    for r := range s {
        ms[s[r] - 'a'] += 1
        if r < len(p) - 1 {
            continue
        }
        if is_equal(ms, mp) {
            res = append(res, l)
        }
        ms[s[l] - 'a'] -= 1
        l++
    }
    
    return res
}

func is_equal(ms, mp [26]byte) bool {
    for i := range mp {
        if ms[i] != mp[i] {
            return false
        }
    }
    return true
}